dlp/deid.py

```diff
@@ -0,0 +1,548 @@
+
+    # Construct FPE configuration dictionary
+    crypto_replace_ffx_fpe_config = {
+        'crypto_key': {
+            'kms_wrapped': {
+                'wrapped_key': wrapped_key,
+                'crypto_key_name': key_name
+            }
+        },
+        'common_alphabet': alphabet
+    }
+
+    # Add surrogate type
+    if surrogate_type:
+        crypto_replace_ffx_fpe_config['surrogate_info_type'] = {
+            'name': surrogate_type
+        }
+
+    # Construct deidentify configuration dictionary
+    deidentify_config = {
+        'info_type_transformations': {
+            'transformations': [
+                {
+                    'primitive_transformation': {
+                        'crypto_replace_ffx_fpe_config':
+                            crypto_replace_ffx_fpe_config
+                    }
+                }
+            ]
+        }
+    }
+
+    # Convert string to item
+    item = {'value': string}
+
+    # Call the API
+    response = dlp.deidentify_content(parent,
+                                      deidentify_config=deidentify_config,
+                                      item=item)
+    # Print results
+    print(response.item.value)
+# [END deidentify_with_fpe]
+
+
+# [START reidentify_with_fpe]
+def reidentify_with_fpe(parent, string, alphabet=None,
+                        surrogate_type=None, key_name=None, wrapped_key=None):
+    """Uses the Data Loss Prevention API to reidentify sensitive data in a
+    string that was encrypted by Format Preserving Encryption (FPE).
+    Args:
+        item: The string to deidentify (will be treated as text).
+        alphabet: The set of characters to replace sensitive ones with. For
+            more information, see https://cloud.google.com/dlp/docs/reference/
+            rest/v2beta2/organizations.deidentifyTemplates#ffxcommonnativealphabet
+        surrogate_type: The name of the surrogate custom info type to used
+            during the encryption process.
+        key_name: The name of the Cloud KMS key used to encrypt ('wrap') the
+            AES-256 key. Example:
+            keyName = 'projects/YOUR_GCLOUD_PROJECT/locations/YOUR_LOCATION/
+            keyRings/YOUR_KEYRING_NAME/cryptoKeys/YOUR_KEY_NAME'
+        wrapped_key: The encrypted ('wrapped') AES-256 key to use. This key
+            should be encrypted using the Cloud KMS key specified by key_name.
+    Returns:
+        None; the response from the API is printed to the terminal.
+    """
+    # Import the client library
+    import google.cloud.dlp
+
+    # Instantiate a client
+    google.cloud.dlp.DlpServiceClient.SERVICE_ADDRESS = 'autopush-dlp.sandbox.googleapis.com'
+    dlp = google.cloud.dlp.DlpServiceClient()
+
+    # Add parent
+    parent = dlp.project_path(parent)
+
+    # Wrapped key can not be base64 encoded
+    import base64
+    wrapped_key = base64.b64decode(wrapped_key)
+
+    # Construct Deidentify Config
+    reidentify_config = {
```

Comment:
this is a bit wild. @andrewsg wdyt about using the real proto types here?

Reply:
Yeah, they're all really heavily nested. What's the advantage of using the real proto types vs. just breaking up the dictionaries into separate steps of construction and adding them together iteratively?

Reply:
It's less about breaking it up into steps and more about knowing the *actual* structure of the item in question. When using the message types directly you get a better sense for how things are constructed and where you can look to find out additional parameters. With just a big o' nested dictionary it's sort of a wild west - the user is left asking "well, how did they know how to construct this? How do *I* know how to construct this?".

Reply:
I agree with this diagnosis of the problem for sure, but I'm not sure about the solution. If we use the real protos here people will just run into the proto help() crash bug. Node doesn't use proto objects; does it have a better user experience for discovering this stuff?

Reply:
Node is also a *vastly* different community with different expectations. Remember, Python is strongly moving towards strongly typed. Seeing giant dictionary with no details on the structure or type constraints isn't wonderful.

Proto help() crash bug has been fixed and should hopefully be released soon.

Reply:
Hmm, okay. I'll buy that. I think that's a bigger change to our focus than we can reasonably do in this PR, though; can we accept this part as-is and do another pass this or next week to see what it looks like with protos instead?